feat: Add support for circular references in schema object

derberg opened the pull request:
- I added dist to the PR because I needed it in the repo so I could test it in my projects. Of course, once you are ok with the PR to merge it I will remove them
- I added tests that prove it works fine, and I also tested it manually with my project and I'm happy with the outcome

Resolves: #113

RomanHotsiy commented on .gitignore:
```diff
@@ -11,7 +11,6 @@ pids
 lib-cov
 
 # dist
-dist/
```
Could you return dist folder back to .gitignore? We do not commit bundled artifacts to source control

derberg replied:
yeap, on it, I removed it and committed dist only because I wanted to use the fork in my projects to manually make sure my changes work as expected

RomanHotsiy commented on src/traverse.js:
```diff
@@ -1,15 +1,27 @@
 
 let $refCache = {};
+//for circular JS references we use additional array and not object as we need to compare entire schemas and not strings
+let circleCache = [];
```
I would rename `circleCache` to `seenSchemasStack`

RomanHotsiy reviewed (changes requested):
Looks good, left a few comments.

derberg commented:
@RomanHotsiy thanks so much for looking into it 🙇 I just made a necessary fixes

RomanHotsiy commented on src/traverse.js:
```diff
@@ -89,6 +97,8 @@ export function traverse(schema, options, spec, context) {
     }
   }
 
+  seenSchemasStack.pop();
```
oh, I think I've found a potential bug. This code can be not reached because of early returns before. So there may be false-positive circular detections. Here is the potential example that may break:

```js
let a = {
  allOf: [ {type: "string"} ]
};

let nonCircularSchema = {
  type: 'object',
  properties: {
    a: a,
    b: a
  }
}
```

I haven't tested it. Could you please verify it doesn't detect false-positive circular refs.

derberg replied:
```
{
        a: 'string',
        b: 'string'
      }
```

Above is returned, as expected, I also added a test case. 
This part of the code is to actually prevent false-positives

RomanHotsiy replied:
@derberg, sorry for the delay

I just checked it. Last time I was in a hurry so the example I sent you works correctly indeed.

But the issue is real. I just verified it. I pushed a commit to this PR where I updated the test case you created so it fails now. It works OK without your changes.

The stack is not `pop`-ed in some cases (because of early returns). Plz, Either refactor the code somehow or add stack pop on every return. (the former would preferable 😄 )

derberg replied:
@RomanHotsiy please have a look now. Popping had to be added before one return. I extended the test case to the maximum to cover all the cases where we have return statements. I think it should be better. It is my first interaction with the project, I have to admit I prefer to add more and better test cases rather than do some refactoring 😄

derberg commented:
@RomanHotsiy I've added a test case to make sure there are no false-positives according to your example

RomanHotsiy commented on src/traverse.js:
```diff
@@ -89,6 +96,7 @@ export function traverse(schema, options, spec, context) {
     }
   }
 
+  seenSchemasStack.pop();
```
I've noticed you push to the stack conditionally (only if there is a context) but pop unconditionally. There must be something wrong with this.

derberg replied:
you were right, good 👁️ , sorry for that

RomanHotsiy commented on test/integration.spec.js:
```diff
@@ -603,4 +603,149 @@ describe('Integration', function() {
+          d: d,
+          dd: d,
+          e: e,
+          ee: e,
```
It shouldn't pass if I read the current code correctly. It must be related to unconditional pop from the stack, see above.

derberg replied:
will push fix in a second

RomanHotsiy reviewed (changes requested):
I've noticed another potential issue there. I wasn't able to find some time to set up a negative test case but I'm pretty sure the code is not fully correct.

Could you please figure it out? This is non-obvious so I don't have a full hour or two to investigate it carefully at the moment.

Sorry for pushing it back-and-forth and for the delays. Thanks for understanding and thanks for your time!

derberg commented:
@RomanHotsiy I solved the issues. I have to agree that it looks strange, the popping before every return, but I'm afraid introducing refactor in this PR will only complicate things.

I'm also now sure about the issue with coveralls. I added needed tests, the utils functions (I added 2 in this PR) are not directly tested so I'm not sure what to do here. And locally when running coverage gulp task, I get 100% coverage report.

RomanHotsiy commented:
@derberg thanks! Great work 👍

derberg commented:
@RomanHotsiy no worries, my pleasure. Thanks a lot for your patience and great review!
